src/hera/workflows/_mixins.py

```diff
@@ -467,62 +467,88 @@ def _build_arguments(self) -> Optional[ModelArguments]:
-                arguments += self._get_deduped_params_from_items(parameter_names, kwargs["with_items"])
-
-            if "source" in kwargs:
-                arguments += _get_source_params_from_kwargs(parameter_names, artifact_names, kwargs["source"], kwargs)
-
-            # it is possible for the user to pass `arguments` via `kwargs` along with `with_param`. The `with_param`
-            # additional parameters are inferred and have to be added to the `kwargs['arguments']` for otherwise
-            # the task will miss adding them when building the final arguments
-            kwargs["arguments"] = arguments
-
             return Task(*args, template=self, **kwargs)
         except InvalidType:
             pass
 
         raise InvalidTemplateCall("Container is not under a Steps, Parallel, or DAG context")
 
+    def _extract_function_argo_arguments(self, **kwargs) -> List:
+        """Returns a list of Parameters and Artifacts extracted from kwargs that belong to the function.
+
+        We inspect the kwargs in a function call, and where assigned values are a Parameter
+        or Artifact, we append them to the arguments list with names replaced. If the
+        function takes a kwarg that name-clashes with any Step or Task field (except "arguments",
+        which are dealt with above), then we raise an error, as `arguments` should be used as a
+        workaround if the function kwarg cannot be renamed.
+
+        Example 1 (artifacts): a kwarg to a function like
+        `message=prev_step.get_artifact("artifact-name")` should
+        pass through a new artifact called "message" with a "from" value of
+        "{{steps.prev-step.outputs.artifacts.artifact-name}}".
+
+        Example 2 (parameters): a kwarg to a function like
+        `message=prev_step.get_parameter("parameter-name")` should
+        pass through a new parameter called "message" with a value of
+        "{{steps.prev-step.outputs.parameters.artifact-name}}".
+
+        Example 3 (basic data types): a kwarg to a function like
+        `message="hello"` should pass through a Parameter with
+        name "message" and value "hello"
+        """
+        from hera.workflows.script import ARGUMENT_TYPE_KWARGS, CLASHING_KWARGS
+
+        arguments = []
+        for arg_name, arg_value in kwargs.items():
+            if isinstance(arg_value, (Parameter, ModelParameter, Artifact, ModelArtifact)):
+                if arg_name not in CLASHING_KWARGS:
+                    # replace names for each argo argument so that this  step/task receives them with correct names
+                    arg_value.name = arg_name
+                    arguments.append(arg_value)
+                elif arg_name not in ARGUMENT_TYPE_KWARGS:
+                    raise ValueError(
+                        f"'{arg_name}' clashes with Step/Task kwargs. Rename '{arg_name}' or "
+                        "pass your Parameter/Artifact through 'arguments' or 'with_param'"
+                    )
```

Comment:
Small piece of feedback, ignore if you disagree. I generally suggest using truthy `is` vs falsy `is not` statements as they seem to be slightly easier to understand:
```suggestion
                if arg_name in CLASHING_KWARGS:
                    raise ValueError(
                        f"'{arg_name}' clashes with Step/Task kwargs. Rename '{arg_name}' or "
                        "pass your Parameter/Artifact through 'arguments' or 'with_param'"
                    )
                elif arg_name in ARGUMENT_TYPE_KWARGS:
                    # replace names for each argo argument so that this  step/task receives them with correct names
                    arg_value.name = arg_name
                    arguments.append(arg_value)
```

Reply:
> arg_value.name = arg_name

Is this replacement safe to do? 🤔 I wonder if this name change affects references downstream of other consumer step/tasks

Reply:
>Is this replacement safe to do? 

I don't believe so, no! Good catch!

Reply:
>slightly easier to understand

I would usually agree, but in this case the if statements in your comment aren't equivalent to mine.

The `if`s are set up in this way because `ARGUMENT_TYPE_KWARGS` is a subset of `CLASHING_KWARGS`.

i.e. this means if we're currently looking at a Parameter/Artifact value (from the `isinstance`):
- we first check the `not in CLASHING_KWARGS` condition as a means of saying "is this a func param?" in which case, full steam ahead.
- Then, if it _is_ `in CLASHING_KWARGS` so it _could_ be in `ARGUMENT_TYPE_KWARGS` which we need to allow (which your code would not). So then we're looking at the `elif arg_name not in ARGUMENT_TYPE_KWARGS`. i.e. we can say the arg is clashing, but we can still allow arguments and with_param usage.

An alternative is a single `if` checking if in clashing_kwargs and not in arg_type_kwargs like so:

```suggestion
                if arg_name in CLASHING_KWARGS and arg_name not in ARGUMENT_TYPE_KWARGS:
                    raise ValueError(
                        f"'{arg_name}' clashes with Step/Task kwargs. Rename '{arg_name}' or "
                        "pass your Parameter/Artifact through 'arguments' or 'with_param'"
                    )
                # replace names for each argo argument so that this  step/task receives them with correct names
                arg_value.name = arg_name
                arguments.append(arg_value)
```

I think this is better actually so will use it. Thanks for your suggestion to help rethink it!

Reply:
Ah, got it, thank you for the discussion!